agent/consul/controller/controller_test.go

```diff
@@ -0,0 +1,228 @@
+package controller
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/hashicorp/consul/agent/consul/fsm"
+	"github.com/hashicorp/consul/agent/consul/state"
+	"github.com/hashicorp/consul/agent/consul/stream"
+	"github.com/hashicorp/consul/agent/structs"
+	"github.com/hashicorp/go-hclog"
+	"github.com/stretchr/testify/require"
+)
+
+func TestBasicController(t *testing.T) {
+	t.Parallel()
+
+	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
+	defer cancel()
+
+	reconciler := newTestReconciler(false)
+
+	publisher := stream.NewEventPublisher(1 * time.Millisecond)
+	go publisher.Run(ctx)
+
+	// get the store through the FSM since the publisher handlers get registered through it
+	store := fsm.NewFromDeps(fsm.Deps{
+		Logger: hclog.New(nil),
+		NewStateStore: func() *state.Store {
+			return state.NewStateStoreWithEventPublisher(nil, publisher)
+		},
+		Publisher: publisher,
+	}).State()
+
+	for i := 0; i < 200; i++ {
+		entryIndex := uint64(i + 1)
+		name := fmt.Sprintf("foo-%d", i)
+		require.NoError(t, store.EnsureConfigEntry(entryIndex, &structs.IngressGatewayConfigEntry{
+			Kind: structs.IngressGateway,
+			Name: name,
+		}))
+	}
+
+	go New(publisher, reconciler).Subscribe(&stream.SubscribeRequest{
+		Topic:   state.EventTopicIngressGateway,
+		Subject: stream.SubjectWildcard,
+	}).WithWorkers(10).Start(ctx)
```

Comment:
@sarahalsmiller @t-eckert I believe this is essentially what we'll need to do on startup for each type we want to control, `EventTopicAPIGateway`, `EventTopicHTTPRoute`, etc.

The more interesting part is just wiring up an implementation of `type Reconciler interface` defined elsewhere in this PR that handles reconciling each type.

Reply:
Exactly, that's what I was fussing with as a "baby controller" for testing.

Reply:
BabyController™